Remove init container and annotation used for Prometheus volume cleanup migration

vicwicker opened the pull request:
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:

We recently identified that certain Prometheus volumes still contain the `prometheus-` folder with outdated Prometheus data. The following PR, with further details, addressed this issue by implementing a cleanup process:

- https://github.com/gardener/gardener/pull/12219

The cleanup code introduced a new init container in the Prometheus resource. For the garden, longterm, aggregate, cache, and seed Prometheus, this init container was added directly to the Prometheus resource. In the case of shoot Prometheus, where successful shoot reconciliation can't be guaranteed, the init container was temporarily added to each shoot Prometheus resource as a one-time migration during the gardenlet startup. Successfully cleaned up shoot Prometheus were annotated to prevent repeated execution of this cleanup process during gardenlet startup.

Now, after three Gardener releases, it is time to drop this init container and annotation from the Prometheus resources and eliminate as much of the migration code as possible.

**Special notes for your reviewer**:

/cc @istvanballok @rfranzke 

**Release note**:

```other operator
Remove the init container and annotation used for the Prometheus volume cleanup migration from Prometheus resources
```

istvanballok reviewed (commented):
We are considering to prepare the commits differently,
- first revert #12219 entirely, and
- then readd the pieces that need to be retained.

That approach will show in the diff the things that "stay", i.e. things that could be reviewed, and not the things that are deleted.
This will allow the reviewer to find things that might be deleted in addition - an aspect that is not obvious in the current representation.

vicwicker committed:
Remove Prometheus annotation from gardenlet startup

This commit introduces a new migration step during the gardenlet startup
to remove the Prometheus annotation used in the previous migration step
to mark it as cleaned up.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

vicwicker committed:
Revert "Clean up obsolete `prometheus-` folder from Prometheus volumes (#12219)"

This reverts commit 3b6c17c808fb30300782408ec4172d90c8e5e5b5. The
Prometheus annotations added in the original commit are not
automatically removed by this revert. There will be a follow up commit
to remove them.

git fetch origin pull/12219/head # d1e7cd5aaf
git fetch origin pull/12728/head # 4b16e434f15
gitk HEAD 4b16e434f15 d1e7cd5aaff ^3b6c17c808^ -- $(git log --oneline -1 --numstat 3b6c17c808 | tail -n +2 | awk '{print $3}')

Conflicts:
  cmd/gardenlet/app/migration.go
    Resolve conflicts manually by removing the function definition and
    function call that was added in the original PR.
  pkg/apis/resources/v1alpha1/types.go
    Manual conflict resolution.
  pkg/component/observability/monitoring/prometheus/component.go
    Manual conflict resolution.
  pkg/component/observability/monitoring/prometheus/prometheus.go
    Manual conflict resolution.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>

istvanballok reviewed (approved):
/lgtm

rfranzke reviewed (commented):
Thank you for following up :)
/approve
